Bug 1740447: Only set resource limits when value is greater then zero

jcantrill opened the pull request:
bug 1740447. Only update the cpu/memory limit if they are greater then zero

https://bugzilla.redhat.com/show_bug.cgi?id=1740447

richm commented on pkg/k8shandler/common_test.go:
```diff
@@ -542,63 +543,17 @@ func buildNoCPULimitResource(cpuRequest, memLimit, memRequest resource.Quantity)
-	}
-
-	return true
+	return printResource(lhs) == printResource(rhs)
```
in general, you are *not* guaranteed the ordering of elements.  Is the ordering guaranteed in this case?  If so, is there a citation?  I don't want to make this code more fragile.

jcantrill replied:
The issue occurs because we are directly comparing stucts and not the value of what is ultimately serialized and ingested by the apiserver.  Though we are not guaranteed of the order, I think the struct is reasonably small and serialization is generally  consistent .  I don't believe we see cases where for instance the `meta` sometimes is serialized after `spec`

richm replied:
It seems backwards that comparing the actual data doesn't work, and we have to serialize to string to do the actual comparison.  Maybe structs work differently than maps/arrays - with maps/arrays you can use `reflect.DeepEqual`

jcantrill replied:
let me investigate. Maybe it was because of the way we displayed failure in the print

jcantrill replied:
I only am able to demonstrate the difference when the struct is serialized.  Unfortunately 0 and nil in this context are not considered to be equivalent

richm replied:
@bparees what's the best way to compare structs like this?

bparees replied:
quantities where you want zero and nil to be treated as equivalent?  i'm not sure we have a generic utility for that.

richm replied:
@bparees then is marshal to string, and compare strings, ok?

@jcantrill it really seems as if you need to write a custom comparison for `Resource` objects.  I know you said ". . .not the value of what is ultimately serialized and ingested by the apiserver." - Yes, the api server is going to receive a serialized string - but the api server is going to unmarshal that into objects and use the objects, not the serialized string.

bparees replied:
why can't we just fix the comparison to handle nil/zero explicitly?

jcantrill replied:
@bparees im not sure I understand what you mean?  This does an equivalent comparison.  @richm the apiserver is going to unmarshal into an object but `ResourceLimits` is a type alias to a maps so no entry means you have nothing for cpu/or memory versus zero.  It's different.

bparees replied:
i mean "if (lhs.Limits.CPU==0 || lhs.Limts.CPU==nil) && (rhs.Limits.CPU==0 || lhs.Limits.CPU==nil) then they are equal"

jcantrill replied:
`reflect.DeepEquals` exposes the issue properly

richm commented on pkg/k8shandler/deployment.go:
```diff
@@ -554,48 +563,63 @@ func (node *deploymentNode) isChanged() bool {
-			nodeContainer.Resources.Requests = v1.ResourceList{}
-		}
+		var updatedContainer v1.Container
+		updatedContainer, changed = updateResources(node, nodeContainer, desiredContainer)
```
If changed was set to `true` by an earlier loop iteration, this may reset it to `false`, will will cause `isChanged` to return the wrong result code.  maybe something like this:
```go
                var updatedContainer v1.Container
                var resourceChanged bool
 		if updatedContainer, resourceChanged = updateResources(node, nodeContainer, desiredContainer); resourceChanged {
                        changed = true
                }
```

richm commented on pkg/k8shandler/deployment.go:
```diff
@@ -554,48 +563,63 @@ func (node *deploymentNode) isChanged() bool {
+
+	// check that both exist
+
+	if nodeContainer.Image != desiredContainer.Image {
```
why are we checking `Image` in `updateResources`?

jcantrill replied:
This was grabbed in the refactor.  I'll just move it out.

richm commented on pkg/k8shandler/deployment.go:
```diff
@@ -554,48 +563,63 @@ func (node *deploymentNode) isChanged() bool {
+	// Check CPU requests
+	if desiredContainer.Resources.Requests.Cpu().Cmp(*nodeContainer.Resources.Requests.Cpu()) != 0 {
+		logrus.Debugf("Resource '%s' has different CPU Request than desired", node.self.Name)
+		nodeContainer.Resources.Requests[v1.ResourceCPU] = *desiredContainer.Resources.Requests.Cpu()
```
we don't need to do the `IsZero` checking for `Requests` cpu and memory?

jcantrill replied:
updated to check requests

richm commented:
why use `logrus` in some places and `logger` in others?

jcantrill commented:
> why use `logrus` in some places and `logger` in others?

The log framework changed when we updated the SDK release.  I added the logger class as an intermediate to initialize `logrus` because it can't be done the same like before the sdk bump (and debug logs are no longer available to our code).  We really need to move completely to the newer framework.  This seems to be a sane first step to me.

ewolinetz commented on pkg/k8shandler/common_test.go:
```diff
@@ -542,63 +543,17 @@ func buildNoCPULimitResource(cpuRequest, memLimit, memRequest resource.Quantity)
-			pretty = fmt.Sprintf("%s\t\tCpu: %s\n", pretty, cpuLimit)
-		}
+	pretty, err := json.MarshalIndent(resource, "", "  ")
+	if err != nil {
```
for the record... this function is only used for making the resources appear in human readable form when we need to notify that values were different. it seems odd that we'd have a potential to fail during what amounts to a `toString`...

jcantrill replied:
Agreed. but its the only way we know Serialization failed

ewolinetz replied:
I think the point I was trying to make was that maybe we shouldn't have used something that could fail here.. i realize this is more streamlined than what we had before, but was there a reason to change it?

jcantrill replied:
It was changed because the previous output didn't properly convey why there was a failure; it was visually provided no difference in exp vs act.  The chance of an error here is likely small given golang is pretty good about Marshalling struct.  I would anticipate we likely never see this error.
